backend/src/halo2_impl.rs

```diff
@@ -21,26 +21,29 @@ impl<T: FieldElement> BackendImpl<T> for Halo2Prover {
         fixed: &[(&str, Vec<T>)],
         witness: &[(&str, Vec<T>)],
         prev_proof: Option<Proof>,
-        output_dir: &Path,
-    ) -> io::Result<()> {
+        output_dir: Option<&Path>,
+    ) -> io::Result<Option<Proof>> {
         let (proof, fname) = match prev_proof {
-            Some(proof) => (self.prove_aggr(pil, fixed, witness, proof), "proof.bin"),
-            None => (self.prove_ast(pil, fixed, witness), "proof_aggr.bin"),
+            Some(proof) => (
+                self.prove_aggr(pil, fixed, witness, proof),
+                "proof_aggr.bin",
+            ),
+            None => (self.prove_ast(pil, fixed, witness), "proof.bin"),
         };
 
-        {
+        if let Some(output_dir) = output_dir {
```

Comment:
Now that the proof is returned, we don't need to write it to disk. In general the only crate which has fs access should be `powdr_cli`.

Reply:
This in turn should simplify the function signatures as `output_dir` shouldn't be required?

Reply:
Me and Leo talked earlier and we considered all that. I think he prefers your way, but hear my reasons:
- In our current model we have an output dir and dump all the compilation artifacts there, with predefined names. The way I propose, we keep this model, and let the backend control the naming of the artifacts it can generate.
- If we are to keep the pilcom as a backend, it needs to know the output dir, because all it does is write files.

Reply:
I think the main culprit here is the pilcom backend. It doesn't return a proof **and** creates a non-proof file that needs to be written to disk in our current workflow.

> In general the only crate which has fs access should be powdr_cli.

I agree with this, but I don't know how to solve it for the pilcom artifact because

> ... So, the natural place for this function to be is in the backend.

I also agree with this.

I would still prefer the proof not to be written here, but that doesn't solve pilcom.